packages/rest/src/http.ts

```diff
@@ -0,0 +1,194 @@
+}
+
+export function getQueryParamName(program: Program, entity: Type) {
+  return program.stateMap(queryFieldsKey).get(entity);
+}
+
+export function isQueryParam(program: Program, entity: Type) {
+  return program.stateMap(queryFieldsKey).has(entity);
+}
+
+const pathFieldsKey = Symbol();
+export function $path(program: Program, entity: Type, paramName: string) {
+  if (!paramName && entity.kind === "ModelProperty") {
+    paramName = entity.name;
+  }
+  program.stateMap(pathFieldsKey).set(entity, paramName);
+}
+
+export function getPathParamName(program: Program, entity: Type) {
+  return program.stateMap(pathFieldsKey).get(entity);
+}
+
+export function isPathParam(program: Program, entity: Type) {
+  return program.stateMap(pathFieldsKey).has(entity);
+}
+
+const bodyFieldsKey = Symbol();
+export function $body(program: Program, entity: Type) {
+  program.stateSet(bodyFieldsKey).add(entity);
+}
+
+export function isBody(program: Program, entity: Type) {
+  return program.stateSet(bodyFieldsKey).has(entity);
+}
+
+export function hasBody(program: Program, parameters: ModelTypeProperty[]): boolean {
+  return parameters.find((p) => isBody(program, p)) !== undefined;
+}
+
+export type HttpVerb = "get" | "put" | "post" | "patch" | "delete";
+
+interface OperationRoute {
+  verb: HttpVerb;
+  subPath?: string;
+}
+
+const operationRoutesKey = Symbol();
+
+function setOperationRoute(program: Program, entity: Type, verb: OperationRoute): void {
+  if (entity.kind === "Operation") {
+    if (!program.stateMap(operationRoutesKey).has(entity)) {
+      program.stateMap(operationRoutesKey).set(entity, verb);
+    } else {
+      reportDiagnostic(program, {
+        code: "http-verb-duplicate",
+        format: { entityName: entity.name },
+        target: entity,
+      });
+    }
+  } else {
+    reportDiagnostic(program, {
+      code: "http-verb-wrong-type",
+      format: { verb: verb.verb, entityKind: entity.kind },
+      target: entity,
+    });
+  }
+}
+
+export function getOperationRoute(program: Program, entity: Type): OperationRoute | undefined {
+  return program.stateMap(operationRoutesKey).get(entity);
+}
+
+export function $get(program: Program, entity: Type, subPath?: string) {
+  setOperationRoute(program, entity, {
+    verb: "get",
+    subPath,
+  });
+}
+
+export function $put(program: Program, entity: Type, subPath?: string) {
+  setOperationRoute(program, entity, {
+    verb: "put",
+    subPath,
+  });
+}
+
+export function $post(program: Program, entity: Type, subPath?: string) {
+  setOperationRoute(program, entity, {
+    verb: "post",
+    subPath,
+  });
+}
+
+export function $patch(program: Program, entity: Type, subPath?: string) {
+  setOperationRoute(program, entity, {
+    verb: "patch",
+    subPath,
+  });
+}
+
+export function $delete(program: Program, entity: Type, subPath?: string) {
+  setOperationRoute(program, entity, {
+    verb: "delete",
+    subPath,
+  });
+}
+
+setDecoratorNamespace(
```

Comment:
I think we talked about this but can you remind me why the `export const namespace` trick doesn't work here?

Reply:
It's because I'm already exporting a `namespace` value for `Cadl.Rest` and if I export more than one `namespace` variable from 2+ modues in this library, there will be a conflict at the main `lib.ts` file.  

Perhaps a better way to add a helper for this would be to loop through all functions in the module scope and add the namespace to all functions that start with `$` so that they don't have to be explicitly referenced in the function call?